core/src/main/java/overflowdb/AdjacentNodes.java

```diff
@@ -10,20 +8,86 @@
   public final Object[] nodesWithEdgeProperties;
 
   /* store the start offset and length into the above `adjacentNodesWithEdgeProperties` array in an interleaved manner,
-   * i.e. each adjacent edge type has two entries in this array. */
-  public final PackedIntArray edgeOffsets;
+   * i.e. each adjacent edge type has two entries in this array.
+   * The type is one of byte[], short[] or int[]*/
+  public final Object offsets;
 
   /**
    * create an empty AdjacentNodes container for a node with @param numberOfDifferentAdjacentTypes
    */
   public AdjacentNodes(int numberOfDifferentAdjacentTypes) {
     nodesWithEdgeProperties = new Object[0];
-    edgeOffsets = PackedIntArray.create(numberOfDifferentAdjacentTypes * 2);
+    offsets = new byte[numberOfDifferentAdjacentTypes * 2];
   }
 
-  public AdjacentNodes(Object[] nodesWithEdgeProperties, PackedIntArray edgeOffsets) {
+
+  public AdjacentNodes(Object[] nodesWithEdgeProperties, Object offsets) {
     this.nodesWithEdgeProperties = nodesWithEdgeProperties;
-    this.edgeOffsets = edgeOffsets;
+    if(offsets == null) throw new RuntimeException("NPE");
+    if(offsets instanceof byte[] | offsets instanceof short[] | offsets instanceof int[])
+      this.offsets = offsets;
+    else throw new RuntimeException();
+  }
+
+  public int getOffset(int pos){
+    if(offsets instanceof byte[]){
+      return ((byte[]) offsets)[pos];
+    } else if(offsets instanceof short[]){
+      return ((short[]) offsets)[pos];
+    } else if (offsets instanceof int[]){
+      return ((int[]) offsets)[pos];
+    } else throw new RuntimeException(offsets.getClass().getName());
+  }
+
+  //returns null if the insertion was in-place; returns a new AdjacentNodes if we need to update
```

Comment:
i really don't like returning `null` as it's setting us up for shooting us in the foot, especially in a `public` method...
1) you could probably make it package-private, i.e. just drop `public`?
2) i may have missed something, but can't we just always `return offsets` here? I didn't go through the callsites in detail but it looked like it wouldn't break them..?

Reply:
:+1: to making everything package private!

Reply:
:+1: